yarn-project/p2p/src/msg_validators/block_proposal_validator/block_proposal_validator.ts

```diff
@@ -24,13 +24,27 @@ export class BlockProposalValidator implements P2PValidator<BlockProposal> {
       }
 
       // Check if transactions are permitted when the proposal contains transaction hashes
-      if (!this.txsPermitted && block.txHashes.length > 0) {
+      const embeddedTxCount = block.txs?.length ?? 0;
+      if (!this.txsPermitted && (block.txHashes.length > 0 || embeddedTxCount > 0)) {
         this.logger.debug(
           `Penalizing peer for block proposal with ${block.txHashes.length} transaction(s) when transactions are not permitted`,
         );
         return PeerErrorSeverity.MidToleranceError;
       }
 
+      // If there are embedded txs, they must be listed in txHashes; if there are no txHashes, there must be no txs
+      const hashSet = new Set(block.txHashes.map(h => h.toString()));
+      if (
+        embeddedTxCount > 0 &&
+        (block.txHashes.length === 0 || block.txs!.some(tx => !hashSet.has(tx.getTxHash().toString())))
```

Comment:
```suggestion
        (block.txs!.some(tx => !hashSet.has(tx.getTxHash().toString())))
```
I think that condition is not needed, right?